Userland/Libraries/LibJS/Runtime/RegExpPrototype.cpp

```diff
@@ -149,38 +149,63 @@ RegexResult RegExpPrototype::do_match(const Regex<ECMA262>& re, const StringView
     return result;
 }
 
-// 22.2.5.2 RegExp.prototype.exec ( string ), https://tc39.es/ecma262/#sec-regexp.prototype.exec
-JS_DEFINE_NATIVE_FUNCTION(RegExpPrototype::exec)
+// 22.2.5.2.1 RegExpExec ( R, S ), https://tc39.es/ecma262/#sec-regexpexec
+Value RegExpPrototype::regexp_exec(GlobalObject& global_object, Object& rx, String const& string)
 {
-    // FIXME: This should try using dynamic properties for 'lastIndex',
-    //        and internal slots [[RegExpMatcher]], [[OriginalFlags]], etc.
+    auto& vm = global_object.vm();
+
+    auto exec = rx.get(vm.names.exec);
+    if (vm.exception())
+        return {};
+
+    if (exec.is_function()) {
+        auto result = vm.call(exec.as_function(), &rx, js_string(vm, string));
+        if (vm.exception())
+            return {};
+
+        if (!result.is_object() && !result.is_null())
+            vm.throw_exception<TypeError>(global_object, ErrorType::NotAnObjectOrNull, result.to_string_without_side_effects());
+
+        return result;
+    }
+
     auto regexp_object = regexp_object_from(vm, global_object);
     if (!regexp_object)
         return {};
 
+    return regexp_builtin_exec(global_object, *regexp_object, string);
+}
+
+// 22.2.5.2.2 RegExpBuiltinExec ( R, S ), https://tc39.es/ecma262/#sec-regexpbuiltinexec
+Value RegExpPrototype::regexp_builtin_exec(GlobalObject& global_object, RegExpObject& rx, String const& string)
+{
+    // FIXME: This should try using dynamic properties for 'lastIndex',
+    //        and internal slots [[RegExpMatcher]], [[OriginalFlags]], etc.
+    auto& vm = global_object.vm();
+
     auto str = vm.argument(0).to_string(global_object);
     if (vm.exception())
         return {};
 
-    StringView str_to_match = str;
+    StringView str_to_match = string;
 
     // RegExps without "global" and "sticky" always start at offset 0.
-    if (!regexp_object->regex().options().has_flag_set((ECMAScriptFlags)regex::AllFlags::Internal_Stateful)) {
-        regexp_object->set(vm.names.lastIndex, Value(0), true);
+    if (!rx.regex().options().has_flag_set((ECMAScriptFlags)regex::AllFlags::Internal_Stateful)) {
```

Comment:
since it's not mandated by the spec I see no reason to not keep the much better `regexp_object` name tbh...